src/backend/distributed/utils/connection_management.c

```diff
@@ -0,0 +1,635 @@
+		if (connection->claimedExclusively)
+		{
+			continue;
+		}
+
+		return connection;
+	}
+
+	return NULL;
+}
+
+
+/*
+ * Return MultiConnection associated with the libpq connection.
+ *
+ * Note that this is comparatively expensive. Should only be used for
+ * backward-compatibility purposes.
+ */
+MultiConnection *
+GetConnectionFromPGconn(struct pg_conn *pqConn)
+{
+	HASH_SEQ_STATUS status;
+	ConnectionHashEntry *entry;
+
+	hash_seq_init(&status, ConnectionHash);
+	while ((entry = (ConnectionHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		dlist_head *connections = entry->connections;
+		dlist_iter iter;
+
+		/* check connection cache for a connection that's not already in use */
+		dlist_foreach(iter, connections)
+		{
+			MultiConnection *connection =
+				dlist_container(MultiConnection, node, iter.cur);
+
+			if (connection->conn == pqConn)
+			{
+				hash_seq_term(&status);
+				return connection;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+
+/*
+ * Close a previously established connection.
+ */
+void
+CloseConnection(MultiConnection *connection)
+{
+	ConnectionHashKey key;
+	bool found;
+
+	/* close connection */
+	PQfinish(connection->conn);
+	connection->conn = NULL;
+
+	strlcpy(key.hostname, connection->hostname, MAX_NODE_LENGTH);
+	key.port = connection->port;
+	strlcpy(key.user, connection->user, NAMEDATALEN);
+	strlcpy(key.database, connection->database, NAMEDATALEN);
+
+	hash_search(ConnectionHash, &key, HASH_FIND, &found);
+
+	if (found)
+	{
+		/* unlink from list */
+		dlist_delete(&connection->node);
+
+		/* we leave the per-host entry alive */
+		pfree(connection);
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("closing untracked connection")));
+	}
+}
+
+
+/*
+ * Close a previously established connection.
+ *
+ * This function closes the MultiConnection associatated with the libpq
+ * connection.
+ *
+ * Note that this is comparatively expensive. Should only be used for
+ * backward-compatibility purposes.
+ */
+void
+CloseConnectionByPGconn(PGconn *pqConn)
+{
+	MultiConnection *connection = GetConnectionFromPGconn(pqConn);
+
+	if (connection)
+	{
+		CloseConnection(connection);
+	}
+	else
+	{
+		ereport(WARNING, (errmsg("could not find connection to close")));
+	}
+}
+
+
+/*
+ * Synchronously finish connection establishment of an individual connection.
+ *
+ * TODO: Replace with variant waiting for multiple connections.
+ */
+void
+FinishConnectionEstablishment(MultiConnection *connection)
+{
+	static int checkIntervalMS = 200;
+
+	/*
+	 * Loop until connection is established, or failed (possibly just timed
+	 * out).
+	 */
+	while (true)
+	{
+		ConnStatusType status = PQstatus(connection->conn);
+		PostgresPollingStatusType pollmode;
+
+		if (status == CONNECTION_OK)
+		{
+			return;
+		}
+
+		/* FIXME: retries? */
+		if (status == CONNECTION_BAD)
+		{
+			return;
+		}
+
+		pollmode = PQconnectPoll(connection->conn);
+
+		/*
+		 * FIXME: Do we want to add transparent retry support here?
+		 */
```

Comment:
I suppose we do have a minor regression compared to ConnectToNode in that regard. I think where it becomes relevant is when we're doing an operation like COPY and making 256 connections, but time out on the 238th one and roll back everything because of it. Would it be hard to add? Wouldn't mind doing it in a separate PR.